### DISCUSSION TITLE
Why is defining a constructor on @ObjectType class forbidden?

### DISCUSSION THREAD

**Discussion by maclockard** [MichalLytek/type-graphql#1201] [Q&A]
The docs call out the fact that defining a constructor on an object type is forbidden:

> Be aware that defining constructors is strictly forbidden and we shouldn't use them there, as TypeGraphQL creates instances of object type classes under the hood by itself.

[source](https://typegraphql.com/docs/types-and-fields.html)

However, I'm not sure I really understand why? I've used constructors for some `@ObjectType` classes and it didn't seem to have an affect.

Additionally, it seems like using a constructor for union types is encouraged: https://github.com/MichalLytek/type-graphql/discussions/1179#discussioncomment-2034034

**Comment by MichalLytek** [comment]
Because object instances are created multiple times in the pipeline and are called without any arguments.

**Reply by antontsvil** [reply]
Can you expand on under what circumstances a type is instantiated without arguments? I tried replicating this scenario and my constructors only ever seem to run once when I call them. Additionally, would switching to a pattern where constructors always null check arguments be sufficient? For example:

```typescript
@InterfaceType()
class HasFoo {
  constructor(args?: Required<HasFoo>) {
    if (args) {
      this.foo = args.foo;
    }
  }

  @Field() foo!: string;
}

@ObjectType({ implements: HasFoo })
class Bar extends HasFoo {
  constructor(args?: Required<Bar & HasFoo>) {
    console.log('Instantiating bar', { args });
    super(args);
    if (args) {
      Object.assign(this, args);
    }
  }

  @Field(type => Int)
  public length!: number;

  @Field(type => ID)
  public name!: string;

  @Field(type => Int)
  public width!: number;
}
```

**Reply by MichalLytek** [reply]
Types are instantiated always when they are used as a resolver method parameter (like `@Root` or `@Arg`).
